scrapy/responsetypes.py

```diff
@@ -86,34 +92,33 @@ def from_filename(self, filename):
         else:
             return Response
 
-    def from_body(self, body):
-        """Try to guess the appropriate response based on the body content.
-        This method is a bit magic and could be improved in the future, but
-        it's not meant to be used except for special cases where response types
-        cannot be guess using more straightforward methods."""
-        chunk = body[:5000]
-        chunk = to_bytes(chunk)
-        if not binary_is_text(chunk):
-            return self.from_mimetype('application/octet-stream')
-        elif b"<html>" in chunk.lower():
-            return self.from_mimetype('text/html')
-        elif b"<?xml" in chunk.lower():
-            return self.from_mimetype('text/xml')
-        else:
-            return self.from_mimetype('text')
-
     def from_args(self, headers=None, url=None, filename=None, body=None):
         """Guess the most appropriate Response class based on
         the given arguments."""
         cls = Response
-        if headers is not None:
+        if cls is Response and body is not None:
+            http_origin = True
+            no_sniff = False
+            content_types = None
+
+            if url and urlparse(url).scheme not in ("http", "https"):
+                http_origin = False
+
+            if headers:
+                if b'Content-Type' in headers:
+                    content_types = (headers[b'Content-Type'],)
+
+                if b'X-Content-Type-Options' in headers and headers[b'X-Content-Type-Options'] == b"nosniff":
+                    no_sniff = True
+
+            mime_type = extract_mime(body, content_types=content_types, http_origin=http_origin, no_sniff=no_sniff)
+            cls = self.from_mimetype(mime_type)
+        if cls is Response and headers is not None:
             cls = self.from_headers(headers)
         if cls is Response and url is not None:
             cls = self.from_filename(url)
         if cls is Response and filename is not None:
             cls = self.from_filename(filename)
```

Comment:
Since the goal is for Scrapy to follow the standard, I believe that it is OK to remove this code completely, and rely on xtractmime only with whatever data is available.